src/main/java/com/aws/greengrass/mqtt/bridge/clients/IoTCoreClient.java

```diff
@@ -109,41 +146,71 @@
-                // If we are unable to unsubscribe, leave the topic in the set so that we can try to remove next time.
+        synchronized (subscribeLock) {
+            if (subscribeFuture != null) {
+                subscribeFuture.cancel(true);
             }
-        });
-
-        Set<String> topicsToSubscribe = new HashSet<>(topics);
-        topicsToSubscribe.removeAll(subscribedIotCoreTopics);
-
-        topicsToSubscribe.forEach(s -> {
-            try {
-                subscribeToIotCore(s);
-                LOGGER.atDebug().kv(TOPIC, s).log("Subscribed to topic");
-                subscribedIotCoreTopics.add(s);
-            } catch (InterruptedException | ExecutionException | TimeoutException e) {
-                LOGGER.atError().kv(TOPIC, s).log("Failed to subscribe");
-            }
-        });
+            this.messageHandler = messageHandler;
+            this.toSubscribeIotCoreTopics = new HashSet<>(topics);
+            LOGGER.atDebug().kv("topics", topics).log("Subscribing to IoT Core topics");
+
+            Set<String> topicsToRemove = new HashSet<>(subscribedIotCoreTopics);
+            topicsToRemove.removeAll(topics);
+            topicsToRemove.forEach(s -> {
+                try {
+                    unsubscribeFromIotCore(s);
+                    LOGGER.atDebug().kv(TOPIC, s).log("Unsubscribed from topic");
+                    subscribedIotCoreTopics.remove(s);
+                } catch (InterruptedException | ExecutionException | TimeoutException e) {
+                    LOGGER.atError().kv(TOPIC, s).setCause(e).log("Unable to unsubscribe");
+                    // If we are unable to unsubscribe, leave the topic in the set
+                    // so that we can try to remove next time.
+                }
+            });
+
+            Set<String> topicsToSubscribe = new HashSet<>(topics);
+            topicsToSubscribe.removeAll(subscribedIotCoreTopics);
+
+            subscribeFuture = executorService.submit(() -> subscribeToTopicsWithRetry(topicsToSubscribe));
+        }
     }
 
     @Override
     public boolean supportsTopicFilters() {
         return true;
     }
 
+    @SuppressWarnings({"PMD.AvoidCatchingGenericException", "PMD.PreserveStackTrace", "PMD.ExceptionAsFlowControl"})
+    private void subscribeToTopicsWithRetry(Set<String> topics) {
+        // retry only if client is connected; skip if offline.
+        // topics left here should be subscribed when the client is back online (onConnectionResumed event)
+        topics.forEach(s -> {
+            try {
+                RetryUtils.runWithRetry(subscribeRetryConfig, () -> {
+                    try {
+                        if (iotMqttClient.connected()) {
+                            subscribeToIotCore(s);
+                            subscribedIotCoreTopics.add(s);
+                        }
+                        // useless return
+                        return null;
+                    } catch (ExecutionException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)